### PR TITLE
Add option to leave out definitions for common modules + namespace models

### PR DESCRIPTION
**Pull request by keithajackson**
When connecting to several LoopBack APIs from an Angular app, multiple definitions of LoopBackAuth and LoopBackResource will 'step on' each other, resulting in unpredictable behavior.

This change keeps the default behavior consistent, but will omit the module block that generates the LoopBackAuth, LoopBackAuthRequestInterceptor, and LoopBackResource components if a fourth parameter (excludeCommonModules) is set to 'true' when generateServices is called.

This will enable loopback-sdk-angular to more seamlessly be used with multiple-API architectures, without having to manually delete sections of code generated by the SDK.

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by keithajackson** [comment]
Rebased to match master. Please let me know if there are any desired changes and I'll be happy to accommodate.

@bajtos

**Comment by bajtos** [comment]
@slnode ok to test

File: lib/services.template.ejs [hunk 1]
```diff
@@ -374,7 +380,9 @@ if (typeof module !== 'undefined' && typeof exports !== 'undefined' &&
         },
       };
     }])
-
+<% } // end if (defineLoopBackAuthRequestInterceptor) %>
+defineLoopBackResource: <%= defineLoopBackResource %>
```
Comment (bajtos) [inline lib/services.template.ejs]: Hmm, what is the purpose of this line?

**Comment by bajtos** [comment]
Hi @keithajackson, thank you for the pull request. Could you please add tests to verify your changes?

If I am reading the code correctly, when `defineLoopBackAuthRequestInterceptor` is true and `defineLoopBackResource` is false, the generated code will miss a semicolon at the end:

```js
module
    // ...
   .config(['$httpProvider', function($httpProvider) {
     $httpProvider.interceptors.push('LoopBackAuthRequestInterceptor');
   }])
   .factory('LoopBackAuthRequestInterceptor', ['$q', 'LoopBackAuth',
     function($q, LoopBackAuth) {
       // ...
     }])
//      ^-- missing semicolon
```

I am proposing to get rid of fine-grained flags `defineLoopBack*` and use a single flag instead.

File: lib/services.js [hunk 2]
```diff
@@ -35,10 +44,23 @@ module.exports = function generateServices(app, ngModuleName, apiUrl) {
     { encoding: 'utf-8' }
   );
 
+  var defineLoopBackAuth = true;
+  var defineLoopBackAuthRequestInterceptor = true;
+  var defineLoopBackResource = true;
+
+  if (excludeCommonModules === "true") {
```
Comment (bajtos) [inline lib/services.js]: Why a string? At this abstraction level, we should expect input arguments in the correct type and let callers to figure out how to convert string input into boolean values.

**Comment by bajtos** [comment]
@keithajackson hello, what's the status of this pull request? Are you still keen to finish it?

**Comment by keithajackson** [comment]
Hi @bajtos - my apologies for letting this go stale - I'm working on writing tests and catching up to master right now.

I'm running into some difficulty getting the test system up and running and was wondering if you might be able to give me a suggestion of what to try next.  I've been following the instructions in `DEVELOPING.md` to run tests with Karma.

When I run `node test.e2e/test-server` in a zsh terminal running Node 5.10.1, I get the output: `Test server is listening on http://localhost:3033`

The second terminal outputs:
```
[2016-06-17 14:41:37.563] [DEBUG] config - Loading config /Users/kjackson/Code/loopback-sdk-angular/karma.conf.js
17 06 2016 14:41:37.578:ERROR [runner]: There is no server listening on port 9876
```

I went into the `karma.conf.js` and changed the `port` from `9876` to `3033` to match the e2e server and re-ran `karma-start`.

The terminal running the test server outputs `POST /run 404 1.274 ms - 17`, but the terminal running `karma start` has another error:

```
[2016-06-17 14:42:38.113] [DEBUG] config - Loading config /Users/kjackson/Code/loopback-sdk-angular/karma.conf.js
Cannot POST /run
```

Does this issue sound familiar? Is there a server or Selenium webdriver I need to be running on port 9876 in addition to the test-server on 3033?

Thank you so much for your time and assistance!

**Comment by bajtos** [comment]
Hey @keithajackson, I am sorry to hear about the difficulties with getting the tests environment set up. Before I dig deeper, I'd like to check whether `npm test` works for you?

**Comment by keithajackson** [comment]
Hi @bajtos ,

I've added tests covering the new features, but  few builds are failing:

1. `default`: when I attempt to access the Jenkins server for details, I get an ERR_ADDRESS_UNREACHABLE error.

2. `loopback-sdk-angular.dependants`:

```
Set build name.
New build name is '#4244 (origin/master)'
Variable with name 'BUILD_DISPLAY_NAME' already exists, current value: '#4244 (origin/master)', new value: '#4244 (origin/master)'
Notifying upstream projects of job completion
Warning: you have no plugins providing access control for builds, so falling back to legacy behavior of permitting any downstream builds to be triggered
Triggering a new build of update-commit-status
Finished: FAILURE
```

3. `pr-builder`:

```
Setting commit status on GitHub for https://github.com/strongloop/loopback-sdk-angular/commit/8d089fc41faf2a6f77abd9d67abc40de45319981
Notifying upstream projects of job completion
Warning: you have no plugins providing access control for builds, so falling back to legacy behavior of permitting any downstream builds to be triggered
Triggering a new build of update-commit-status
Finished: FAILURE
```

4. `dependent-loopback-workspace`:
```
  1) end-to-end empty-server template without explorer "before all" hook: installSandboxPackages:
     Error: timeout of 300000ms exceeded. Ensure the done() callback is being called in this test.
      at [object Object].<anonymous> (node_modules/loopback/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31)

  2) end-to-end api-server template "before all" hook:
     Error: ENOTEMPTY, rmdir '/mnt/jenkins/workspace/loopback-workspace/1de32925/test/sandbox/node_modules/eslint/node_modules/es6-map/node_modules/es5-ext/test/number'
      at Error (native)

  3) end-to-end notes template "before all" hook: installSandboxPackages:
     Error: timeout of 300000ms exceeded. Ensure the done() callback is being called in this test.
      at [object Object].<anonymous> (node_modules/loopback/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31)

  4) end-to-end hello-world template "before all" hook:
     Error: ENOTEMPTY, rmdir '/mnt/jenkins/workspace/loopback-workspace/1de32925/test/sandbox/node_modules/eslint/node_modules/escope/node_modules/es6-weak-map/node_modules/es5-ext/math'
      at Error (native)

  5) end-to-end autoupdate "before all" hook: installSandboxPackages:
     Error: timeout of 300000ms exceeded. Ensure the done() callback is being called in this test.
      at [object Object].<anonymous> (node_modules/loopback/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31)

  6) end-to-end discovery "before all" hook:
     Error: ENOTEMPTY, rmdir '/mnt/jenkins/workspace/loopback-workspace/1de32925/test/sandbox/node_modules/eslint/node_modules/escope/node_modules/es6-weak-map/node_modules/es5-ext/test/array/#'
      at Error (native)

  7) end-to-end testConnection "before all" hook: installSandboxPackages:
     Error: timeout of 300000ms exceeded. Ensure the done() callback is being called in this test.
      at [object Object].<anonymous> (node_modules/loopback/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31)

  8) end-to-end start/stop/restart "before all" hook:
     Error: ENOTEMPTY, rmdir '/mnt/jenkins/workspace/loopback-workspace/1de32925/test/sandbox/node_modules/eslint/node_modules/escope/node_modules/es6-weak-map/node_modules/es5-ext/math/asinh'
      at Error (native)
```

Do you have any recommendations for how to address these issues?
Thanks!

File: lib/services.js [hunk 3]
```diff
@@ -50,10 +50,14 @@ module.exports = function generateServices(app, options) {
       apiUrl: arguments[2],
     };
   }
+  var oldOptions = options;
```
Comment (bajtos) [inline lib/services.js]: AFAICT, this is unused - please remove.

File: lib/services.template.ejs [hunk 4]
```diff
@@ -37,8 +37,6 @@ if (typeof module !== 'undefined' && typeof exports !== 'undefined' &&
 
 <% for (var modelName in models) {
      var meta = models[modelName];
-     // capitalize the model name
-     modelName = modelName[0].toUpperCase() + modelName.slice(1);
```
Comment (bajtos) [inline lib/services.template.ejs]: `modelName` won't be available any more, are you sure it's not used anywhere in the template?

File: lib/services.template.ejs [hunk 5]
```diff
@@ -390,7 +389,6 @@ if (typeof module !== 'undefined' && typeof exports !== 'undefined' &&
         },
       };
     }])
-
```
Comment (bajtos) [inline lib/services.template.ejs]: Looks like unrelated whitespace change, please revert.

File: test.e2e/given.js [hunk 6]
```diff
@@ -58,7 +58,9 @@ define(['angular', 'angularMocks', 'angularResource'], function(angular) {
    * ```
    */
   given.servicesForLoopBackApp = function(options, cb) {
-    options.name = generateUniqueServiceName(options.name);
+    if (! options.name) {
```
Comment (bajtos) [inline test.e2e/given.js]: style:

```js
if (!options.name) {
```

File: test.e2e/test-server.js [hunk 7]
```diff
@@ -105,19 +104,8 @@ masterApp.post('/setup', function(req, res, next) {
...
-        servicesScript = generator.services(lbApp, {
-          ngModuleName: name,
-          apiUrl: apiUrl,
-          includeSchema: includeSchema,
-        });
-      } else {
-        // the old API, test it to verify backwards compatibility
-        servicesScript = generator.services(lbApp, name, apiUrl);
-      }
+      servicesScript = generateService(generator, lbApp, apiUrl, opts);
```
Comment (bajtos) [inline test.e2e/test-server.js]: Please keep "the old API" branch, we need to keep it covered by unit-tests to prevent regressions.

Comment (bajtos) [inline test.e2e/test-server.js]: See https://github.com/strongloop/loopback-sdk-angular/pull/224#discussion_r72442921

Comment (bajtos) [inline test.e2e/test-server.js]: Oh, I see, this is handled by `generateService`, ignore my comment then.

File: test.e2e/test-server.js [hunk 8]
```diff
@@ -161,6 +149,27 @@ masterApp.listen(port, function() {
     runAndExit(process.argv[2], process.argv.slice(3));
 });
 
+function generateService(generator, lbApp, apiUrl, opts) {
+  var servicesScript;
+  if (typeof opts.includeSchema !== 'undefined' ||
```
Comment (bajtos) [inline test.e2e/test-server.js]: In Loopback, we use simpler `opts.includeSchema !== undefined`. Please simplify here and on the lines below too.

File: test.e2e/test-server.js [hunk 9]
```diff
@@ -161,6 +149,27 @@ masterApp.listen(port, function() {
...
+      typeof opts.includeCommonModules !== 'undefined' ||
+      typeof opts.namespaceModels !== 'undefined' ||
+      typeof opts.namespaceDelimiter !== 'undefined') {
+    // the new options-based API
+
+    // build options object for new options-based API
+    var generatorOptions = opts;
+    generatorOptions.ngModuleName = opts.name;
```
Comment (bajtos) [inline test.e2e/test-server.js]: I think you are missing `apiUrl` here?

**Comment by bajtos** [comment]
Hey @keithajackson, sorry for the long delay. I reviewed your changes, they looks mostly good to me. Please address the comments I left above.

As for failing CI builds, I am afraid our CI system is still not stable and still not visible to public. Just make sure `npm test` passes on your local machine, we will check CI results ourselves to see if they are related to this patch.

**Comment by slnode** [comment]
Can one of the admins verify this patch?

File: lib/services.template.ejs (hunk 4 above)
Comment (keithajackson) [inline lib/services.template.ejs]: `modelName` should be available inside the for loop, as this deleted code just updates the variable, rather than creating a new one (the formatting that used to be in the template is now in `getFormattedModelName` in `services.js`).  I double-checked that there isn't anywhere outside the for loop where modelName is used - it's just inside this for loop that we reference it directly.

File: test.e2e/test-server.js (hunk 9 above)
Comment (keithajackson) [inline test.e2e/test-server.js]: Ah yes! Added - thanks!

File: test.e2e/test-server.js (hunk 8 above)
Comment (keithajackson) [inline test.e2e/test-server.js]: Done!

File: lib/services.template.ejs (hunk 5 above)
Comment (keithajackson) [inline lib/services.template.ejs]: Got it!

File: test.e2e/given.js (hunk 6 above)
Comment (keithajackson) [inline test.e2e/given.js]: Sounds good!

File: lib/services.js (hunk 3 above)
Comment (keithajackson) [inline lib/services.js]: Ah yes - removed!

**Comment by keithajackson** [comment]
@bajtos All comments adressed and code rebased against master - thanks so much for close review! Let me know if you have any other comments you'd like addressed or run into any surprises.

I double-checked test after the update+rebase and it all looks good on my machine.

**Comment by bajtos** [comment]
Thank you for the quick update, @keithajackson!

I run `node apidocs/describe-builtin-models.js` using `master` version and the version from your patch. When I compared the outcome, there is one extra empty line:

```diff
--- apidocs/loopback-core-master.js	2016-07-28 14:43:03.000000000 +0200
+++ apidocs/loopback-core.js	2016-07-28 14:43:18.000000000 +0200
@@ -6993,6 +6993,7 @@
       }]);


+
   module
   .factory('LoopBackAuth', function() {
     var props = ['accessTokenId', 'currentUserId', 'rememberMe'];
```

What's your opinion? Do you feel like fixing this small detail, or should we ignore it?

Other than that, the patch LGTM.

Could you please squash all your commits into a single one and provide a descriptive commit message following the [50/72 rule](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html)?

**Comment by keithajackson** [comment]
@bajtos Fixed! I was able to find the blank line causing the documentation inconsistency and removed it.  I've also squashed commits to match the commit style.

Thanks again for your time and thoughtful comments. Let me know if there is anything else I can improve or help with!

**Comment by bajtos** [comment]
@keithajackson awesome! I have landed the patch, thank you for this great contribution 🎉 👏

**Comment by bajtos** [comment]
Released as `loopback-sdk-angular@1.9.0`, enjoy 😄
